chevah/python-package
# [#5281] Added bcrypt and Alpine 3.10.

Scope
=====

Add bcrypt to python-package

This is needed to support private SSH keys encrypted using bcrypt


Changes
=======

Added `bcrypt` package and a simple test for it.

But not on HP-UX, where it is totally unsupported, see comment for details.

On Solaris 10 x64 there were a couple of issues:

  1. System `libffi` is missing `libffi.pc`, which is needed for `bcrypt` build.
  2. With our own `libffi`, `bcrypt` build still fails with:

```
~/runtime/build-solaris10-x64/slave/python-package-solaris-10/build/build/bcrypt-3.1.7 ~/runtime/build-solaris10-x64/slave/python-package-solaris-10/build
"_configtest.c", line 1: warning: implicit function declaration: __sync_synchronize
ld: fatal: file _configtest.o: wrong ELF class: ELFCLASS64
ld: fatal: file processing errors. No output written to _configtest
Note: will not use '__sync_synchronize()' in the C code
***** The above error message can be safely ignored.

"c/_cffi_backend.c", line 7205: warning: assignment type mismatch:
	pointer to void "=" pointer to function(char, char) returning char
"c/_cffi_backend.c", line 7206: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int, long) returning long
"c/_cffi_backend.c", line 7207: warning: assignment type mismatch:
	pointer to void "=" pointer to function(long long, long long) returning long long
"c/_cffi_backend.c", line 7208: warning: assignment type mismatch:
	pointer to void "=" pointer to function(float, double) returning double
"c/_cffi_backend.c", line 7209: warning: assignment type mismatch:
	pointer to void "=" pointer to function(float, double) returning float
"c/_cffi_backend.c", line 7210: warning: assignment type mismatch:
	pointer to void "=" pointer to function(void) returning void
"c/_cffi_backend.c", line 7211: warning: assignment type mismatch:
	pointer to void "=" pointer to function(pointer to int) returning pointer to int
"c/_cffi_backend.c", line 7212: warning: assignment type mismatch:
	pointer to void "=" pointer to function(struct _testfunc7_s {unsigned char a1, short a2}) returning short
"c/_cffi_backend.c", line 7214: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int, ...) returning int
"c/_cffi_backend.c", line 7215: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc7_s {unsigned char a1, short a2}
"c/_cffi_backend.c", line 7216: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc11_s {int a1, int a2}
"c/_cffi_backend.c", line 7217: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc12_s {double a1}
"c/_cffi_backend.c", line 7218: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc13_s {int a1, int a2, int a3}
"c/_cffi_backend.c", line 7219: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc14_s {float a1}
"c/_cffi_backend.c", line 7220: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc15_s {float a1, int a2}
"c/_cffi_backend.c", line 7221: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc16_s {float a1, float a2}
"c/_cffi_backend.c", line 7222: warning: assignment type mismatch:
	pointer to void "=" pointer to function(int) returning struct _testfunc17_s {int a1, float a2}
"c/_cffi_backend.c", line 7223: warning: assignment type mismatch:
	pointer to void "=" pointer to function(pointer to struct _testfunc17_s {int a1, float a2}) returning int
"c/_cffi_backend.c", line 7224: warning: assignment type mismatch:
	pointer to void "=" pointer to function(long double, int) returning long double
"c/_cffi_backend.c", line 7225: warning: assignment type mismatch:
	pointer to void "=" pointer to function(pointer to struct _testfunc7_s {unsigned char a1, short a2}) returning short
"c/_cffi_backend.c", line 7226: warning: assignment type mismatch:
	pointer to void "=" pointer to function(struct _testfunc21_s {int a, int b, int c, int d, int e, int f, int g, int h, int i, int j}) returning int
"c/_cffi_backend.c", line 7227: warning: assignment type mismatch:
	pointer to void "=" pointer to function(struct _testfunc22_s {array[10] of int a}, struct _testfunc22_s {array[10] of int a}) returning struct _testfunc22_s {array[10] of int a}
"c/_cffi_backend.c", line 7228: warning: assignment type mismatch:
	pointer to void "=" pointer to function(pointer to char) returning int
"c/_cffi_backend.c", line 7499: warning: initialization type mismatch
"c/_cffi_backend.c", line 7500: warning: initialization type mismatch
"c/_cffi_backend.c", line 7501: warning: initialization type mismatch
"c/_cffi_backend.c", line 7502: warning: initialization type mismatch
"c/_cffi_backend.c", line 7503: warning: initialization type mismatch
"c/_cffi_backend.c", line 7504: warning: initialization type mismatch
"c/_cffi_backend.c", line 7505: warning: initialization type mismatch
"c/_cffi_backend.c", line 7506: warning: initialization type mismatch
"c/_cffi_backend.c", line 7507: warning: initialization type mismatch
"c/_cffi_backend.c", line 7508: warning: initialization type mismatch
"c/_cffi_backend.c", line 7509: warning: initialization type mismatch
"c/_cffi_backend.c", line 7510: warning: initialization type mismatch
"c/_cffi_backend.c", line 7511: warning: initialization type mismatch
"c/_cffi_backend.c", line 7512: warning: initialization type mismatch
"c/_cffi_backend.c", line 7513: warning: initialization type mismatch
"c/_cffi_backend.c", line 7514: warning: initialization type mismatch
"c/_cffi_backend.c", line 7515: warning: initialization type mismatch
"c/_cffi_backend.c", line 7516: warning: initialization type mismatch
"c/_cffi_backend.c", line 7517: warning: initialization type mismatch
"c/_cffi_backend.c", line 7518: warning: initialization type mismatch
"c/_cffi_backend.c", line 7519: warning: initialization type mismatch
"c/_cffi_backend.c", line 7520: warning: initialization type mismatch
"c/_cffi_backend.c", line 7521: warning: initialization type mismatch
"c/_cffi_backend.c", line 7522: warning: initialization type mismatch
"c/_cffi_backend.c", line 7523: warning: initialization type mismatch
"c/_cffi_backend.c", line 7524: warning: initialization type mismatch
"c/_cffi_backend.c", line 7525: warning: initialization type mismatch
"c/realize_c_type.c", line 653: warning: syntax error:  empty declaration
"c/cdlopen.c", line 196: warning: assignment type mismatch:
	pointer to void "=" pointer to function(pointer to struct _cffi_getconst_s {unsigned long long value, pointer to const struct _cffi_type_context_s {..} ctx, int gindex}) returning int
```

This is fixable with forcing a 32bit build for Solaris 10 on amd64 arch as well. Note that for all the other Unix flavours we are building 32bit binaries, Solaris 10 x64 was the last one standing on 64bit ground.

64bit builds might be fixable on Solaris 10 x64, but I haven't tried hard, not sure it's worth the trouble.

**Drive-by fixes**:
  * support for Alpine 3.10
  * external deps doc fixes and updates.

How to try and test the changes
===============================

reviewers: @adiroiban 

Please review changes in our code: `git diff master chevah_build python-modules/chevah-python-test`

Run the automated tests.

Assess if building 32bit binaries on Solaris 10 amd64 is acceptable or should we try harder to keep the Solaris amd64 builds 64bit-only.
